numpy/_core/src/multiarray/array_converter.c

```diff
@@ -0,0 +1,471 @@
+    }
+
+    int ndescrs = 0;
+    int nDTypes = 0;
+    creation_item *item = self->items;
+    for (int i = 0; i < self->narrs; i++, item++) {
+        DTypes[nDTypes] = item->DType;
+        nDTypes++;
+        if (item->descr != NULL) {
+            descrs[ndescrs] = item->descr;
+            ndescrs++;
+        }
+    }
+
+    if (ensure_inexact) {
+        if (dt_info.dtype != NULL) {
+            PyErr_SetString(PyExc_TypeError,
+                    "extra_dtype and ensure_inexact are mutually exclusive.");
+            goto finish;
+        }
+        Py_INCREF(&PyArray_PyFloatAbstractDType);
+        dt_info.dtype = &PyArray_PyFloatAbstractDType;
+    }
+
+    if (dt_info.dtype != NULL) {
+        DTypes[nDTypes] = dt_info.dtype;
+        nDTypes++;
+    }
+    if (dt_info.descr != NULL) {
+        descrs[ndescrs] = dt_info.descr;
+        ndescrs++;
+    }
+
+    PyArray_DTypeMeta *common_dtype = PyArray_PromoteDTypeSequence(
+            nDTypes, DTypes);
+    if (common_dtype == NULL) {
+        goto finish;
+    }
+    if (ndescrs == 0) {
+        result = NPY_DT_CALL_default_descr(common_dtype);
+    }
+    else {
+        result = PyArray_CastToDTypeAndPromoteDescriptors(
+                ndescrs, descrs, common_dtype);
+    }
+    Py_DECREF(common_dtype);
+
+  finish:
+    Py_XDECREF(dt_info.descr);
+    Py_XDECREF(dt_info.dtype);
+    PyMem_Free(DTypes_and_descrs);
+    return (PyObject *)result;
+}
+
+
+static PyGetSetDef array_converter_getsets[] = {
+    {"scalar_input",
+        (getter)array_converter_get_scalar_input,
+        NULL,
+        NULL, NULL},
+    {NULL, NULL, NULL, NULL, NULL},
+};
+
+
+static PyMethodDef array_converter_methods[] = {
+    {"as_arrays", 
+        (PyCFunction)array_converter_as_arrays,
+        METH_FASTCALL | METH_KEYWORDS, NULL},
+    {"result_type",
+        (PyCFunction)array_converter_result_type,
+        METH_FASTCALL | METH_KEYWORDS, NULL},
+    {"wrap",
+        (PyCFunction)array_converter_wrap,
+        METH_FASTCALL | METH_KEYWORDS, NULL},
+    {NULL, NULL, 0, NULL}
+};
+
+
+static void
+array_converter_dealloc(PyArrayArrayConverterObject *self)
+{
+    creation_item *item = self->items;
+    for (int i = 0; i < self->narrs; i++, item++) {
+        Py_XDECREF(item->array);
+        Py_XDECREF(item->object);
+        Py_XDECREF(item->DType);
+        Py_XDECREF(item->descr);
+    }
+
+    Py_XDECREF(self->wrap);
+    Py_XDECREF(self->wrap_type);
+    PyObject_Del((PyObject *)self);
+}
+
+
+static Py_ssize_t
+array_converter_length(PyArrayArrayConverterObject *self)
+{
+    return self->narrs;
+}
+
+
+static PyObject *
+array_converter_item(PyArrayArrayConverterObject *self, Py_ssize_t item)
+{
+    /* Python ensures no negative indices (and probably the below also) */
+    if (item < 0 || item >= self->narrs) {
+        PyErr_SetString(PyExc_IndexError, "invalid index");
+        return NULL;
+    }
+
+    /* When we only have scalars, the default is to return an array: */
```

Comment:
I was again a bit confused about this, maybe it helps to say "We follow the default setting for as_arrays of CONVERT_IF_NO_ARRAY"